ASTImporter InjectedClassNameType related fixes

balazske opened the pull request:
Provides fix for #346.

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2201,28 +2201,29 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
-              ToDescribed->getInjectedClassNameSpecialization());
-          // Find the injected class name itself and update the type too.
-          CXXRecordDecl *Injected = [&] {
+          auto Injected = [&]()->CXXRecordDecl * {
```
Could you please list all the variables which are taken as reference in the capture list? That makes it cleaner to see which variables could be changed by the lambda.

Perhaps we should name the lambda to make its intention even cleaner. E.g. `auto FindInjectedClass = ...; auto Injected = FindInjectedClass()`

Also, formatting is a bit off here, I suppose it should be `[&]() -> CXXRecordDecl *`

Xazax-hun commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -2543,5 +2543,60 @@ long b = strtol(&a);
   EXPECT_EQ(DeclCounter<UsingShadowDecl>().match(ToTU, usingShadowDecl()), 1u);
 }
 
+TEST_F(Fixture, MakeInjectedClassNameTypeWithPrevDecl) {
```
I wonder if unit test is the best way to to trigger this issue. What about an ASTMerge regression test?

balazske replied:
I could not easily produce a more simple way to get the original assertion.

balazske replied:
What is the difference between ASTMerge and Import tests (where to put a new test)?

martong replied:
The test cases `check-clang-astmerge` are using the Importer directly in a intuitive way (See `lib/frontend/ASTMerge.cpp`). I'd suggest to write a test with these if you really want lit tests. For me the unit tests we have is perfectly fine.

 The other lit tests are in `check-clang-import` which uses the `clang-import-test` binary. And this uses the Importer indirectly through `ExternalASTMerger`. If I understand correctly then this tool does allow to import incomplete Decls and DeclContexts, while the other (`check-clang-astmerge`) does not. 
So we can merge these two files:
```
enum E {
  a = 1,
  b = 2
};
```
```
void expr() {
  static_assert(E::a + E::b == 3);
}
```
However with the first tool `clang -ast-merge` that would not be possible because the second TU would not compile (since E has a missing definition).

Xazax-hun commented on lib/AST/ASTImporter.cpp:
```diff
@@ -2201,28 +2201,29 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
               }
             }
-            assert(false && "No injected-class-name found!");
+            return nullptr;
```
When will this return a nullptr? In case this only reached for non-complete definitions, maybe we could assert that.

balazske replied:
In nullptr case the injected-class-name is not yet imported. But it should be imported later and then the already created InjectedClassNameType is used.

balazske commented:
Code was updated, now there is a ASTMerge test instead of unit test (the test fails without the correction with "Assertion `isa<InjectedClassNameType>(Decl->TypeForDecl)' failed"). Lambda function was removed.

martong commented on test/ASTMerge/injected-class-name-decl-1/test.cpp:
```diff
@@ -0,0 +1,6 @@
+// with InjectedClassNameType there is a PrevDecl of the record.
+
+// RUN: %clang_cc1 -std=c++1z -emit-pch -o %t.ast %S/Inputs/inject1.cpp
+// RUN: %clang_cc1 -std=c++1z -emit-obj -o /dev/null -ast-merge %t.ast %S/Inputs/inject2.cpp
```
Could we remove `Inputs/inject2.cpp` and instead use `test.cpp` as the "to" context:
```
// RUN: %clang_cc1 -std=c++1z -emit-obj -o /dev/null -ast-merge %t.ast %s
```
?

Xazax-hun replied:
Do we need the emit-obj? Can't we just add -fsyntax-only or something like this?

balazske commented:
Code updated: Removed inject2.cpp from test and simplified test.cpp.
(I do not like force push: The history of changes could be visible if old commits remain on the branch.)

Xazax-hun commented:
The advantage of squashing comments and force push is that it will be easier for these to be ported to the newer versions of clang.

balazske commented:
There is a "squash commits" option in GitHub that should do a similar thing but the other commits are not lost (and no bad references are appearing in pull requests).

balazske commented:
Related Phabricator item:
https://reviews.llvm.org/D47450
